cl_manycore/libraries/bsg_manycore_cuda.cpp

```diff
@@ -1938,10 +2019,417 @@ int hb_mc_tile_set_grid_dim_symbols (	hb_mc_manycore_t *mc,
+				hb_mc_coordinate_get_y(tiles[tile_id]), 
+				argc);
+
+
+		error = hb_mc_tile_set_argv_ptr_symbol(	device->mc, map, 
+							device->program->bin,
+							device->program->bin_size,
+							&(tiles[tile_id]),
+							&args_eva);
+		if (error != HB_MC_SUCCESS) { 
+			bsg_pr_err(	"%s: failed to write argv pointer to tile (%d,%d).\n",
+					__func__,
+					hb_mc_coordinate_get_x(tiles[tile_id]),
+					hb_mc_coordinate_get_y(tiles[tile_id]));
+			return error;
+		}
+		bsg_pr_dbg(	"%s: Setting tile (%d,%d) cuda_argv_ptr symbol to 0x%08" PRIx32 ".\n",
+				__func__,
+				hb_mc_coordinate_get_x(tiles[tile_id]),
+				hb_mc_coordinate_get_y(tiles[tile_id]),
+				args_eva);
+
+
+		hb_mc_eva_t finish_signal_eva;
+		size_t sz; 
+		error = hb_mc_npa_to_eva (cfg, map, &(tiles[tile_id]), &(finish_signal_npa), &finish_signal_eva, &sz); 
+		if (error != HB_MC_SUCCESS) { 
+			bsg_pr_err("%s: failed to acquire finish signal address eva from npa.\n", __func__); 
+			return error;
+		}
+
+		error = hb_mc_tile_set_finish_signal_addr_symbol(	device->mc, map, 
+									device->program->bin,
+									device->program->bin_size,
+									&(tiles[tile_id]),
+									&finish_signal_eva);
+		if (error != HB_MC_SUCCESS) {
+			bsg_pr_err(	"%s: failed to write finish signal address to tile (%d,%d).\n",
+					__func__,
+					hb_mc_coordinate_get_x(tiles[tile_id]),
+					hb_mc_coordinate_get_y(tiles[tile_id]));
+			return error;
+		}
+		bsg_pr_dbg(	"%s: Setting tile (%d,%d) cuda_finish_signal_addr symbol to 0x%08" PRIx32 ".\n",
+				__func__,
+				hb_mc_coordinate_get_x(tiles[tile_id]),
+				hb_mc_coordinate_get_y(tiles[tile_id]),
+				finish_signal_eva);
+
+
+		error = hb_mc_tile_set_kernel_ptr_symbol(	device->mc, map, 
+								device->program->bin,
+								device->program->bin_size,
+								&(tiles[tile_id]),
+								&kernel_eva);
+		if (error != HB_MC_SUCCESS) {
+		bsg_pr_err(	"%s: failed to write kernel pointer to tile (%d,%d).\n",
+					__func__,
+					hb_mc_coordinate_get_x(tiles[tile_id]),
+					hb_mc_coordinate_get_y(tiles[tile_id]));
+			return error;
+		}
+		bsg_pr_dbg(	"%s: Setting tile (%d,%d) cuda_kernel_ptr symbol to 0x%08" PRIx32 ".\n",
+				__func__,
+				hb_mc_coordiante_get_x(tiles[tile_id]),
+				hb_mc_coordinate_get_y(tiles[tile_id]),
+				kernel_eva); 
+	}
+
+	return HB_MC_SUCCESS;
+}
 
 
 
 
 
+
+/*! 
+ * Sets a Vanilla Core's cuda_kernel_ptr symbol.
+ * Behavior is undefined if #mc is not initialized with hb_mc_manycore_init().
+ * @param[in] mc         A manycore instance initialized with hb_mc_manycore_init().
+ * @param[in] map        Eva to npa mapping. 
+ * @param[in] bin        Binary elf file. 
+ * @param[in] bin_size   Size of binary file. 
+ * @param[in] coord      Tile coordinates to set the tile group id of.
+ * @param[in] kernel_eva EVA address of the kernel 
+ * @return HB_MC_SUCCESS if successful, otherwise an error code is returned. 
+ */
+static int hb_mc_tile_set_kernel_ptr_symbol (	hb_mc_manycore_t *mc,
```

Comment:
This function and the four or so that follow are very similar. In each one, you are looking up a symbol and setting it to a 32-bit value. I recommend making a base function with a signature like this:

```c
static int hb_mc_tile_set_symbol_value( /* mc, map, bin, bin_size, coord, */ const char *symbol, const uint32_t *v);
```

and then implement your other functions like this:
```c
static int hb_mc_tile_set_kernel_ptr_symbol(hb_mc_manycore_t *mc,hb_mc_eva_map_t *map,unsigned char* bin, size_t bin_size, const hb_mc_coordinate_t *coord,const hb_mc_eva_t *kernel_eva) {
    return hb_mc_tile_set_symbol_value(mc, map, bin, bin_size, coord, "cuda_kernel_ptr", kernel_eva);
}
```

This will reduce the amount of spaghetti code.

Why is this important? Well... imagine you mess something up in ONE of these functions. Odds are, if they're all pretty much copy and paste, you'll have to go back and make the same change 4+ places. But if you do what I've described above, you won't run into this problem.

Reply:
Make a lot of sense! done, thanks!